cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/SystemProgramManagementService.java

```diff
@@ -0,0 +1,211 @@
+    if (programsEnabled.get() == null) {
+      LOG.debug("Programs to run not yet set");
+      return;
+    }
+    try {
+      reconcilePrograms();
+    } catch (Exception ex) {
+      //catch here so a single iteration doesn't stop the service
+      LOG.error("Error in iteration of reconciling programs", ex);
+    }
+  }
+
+  @Override
+  protected ScheduledExecutorService executor() {
+    executor = Executors
+      .newSingleThreadScheduledExecutor(Threads.createDaemonThreadFactory(this.getClass().getSimpleName()));
+    return executor;
+  }
+
+  @Override
+  protected void shutDown() throws Exception {
+    if (executor != null) {
+      executor.shutdown();
+    }
+  }
+
+  private void reconcilePrograms() throws Exception {
+    //take a copy
+    Map<ProgramId, Arguments> enabledProgramsMap = this.programsEnabled.get();
+    Set<ProgramId> enabledPrograms = enabledProgramsMap.keySet();
+    Set<ProgramRunId> programsToStop = new HashSet<>();
+    Set<ProgramId> programsRunning = new HashSet<>();
+    Map<ProgramId, List<ProgramRuntimeService.RuntimeInfo>> programRunsMap = new HashMap<>();
+    Map<ProgramId, RunRecordDetail> programLatestRunMap = new HashMap<>();
+    //Get all current runs
+    List<ProgramRuntimeService.RuntimeInfo> runtimeInfos = programRuntimeService.listAll(ProgramType.values());
+    //Find programs to delete and already running programs
+    for (ProgramRuntimeService.RuntimeInfo runtimeInfo : runtimeInfos) {
+      ProgramId programId = runtimeInfo.getProgramId();
+      if (!programId.getNamespaceId().equals(NamespaceId.SYSTEM)) {
+        //We care only about system programs
+        continue;
+      }
+      if (!enabledPrograms.contains(programId)) {
+        programsToStop.add(runtimeInfo.getController().getProgramRunId());
+      } else {
+        programsRunning.add(programId);
+        //gather run records to stop extra runs for each program
+        gatherRunRecords(programRunsMap, programLatestRunMap, runtimeInfo, programId);
+      }
+    }
+    //find programs to start
+    Set<ProgramId> programsToStart = enabledPrograms.stream()
+      .filter(programId -> !programsRunning.contains(programId))
+      .collect(Collectors.toSet());
+    //start programs
+    startPrograms(programsToStart, enabledProgramsMap);
+    //stop programs
+    programsToStop.forEach(this::stopProgram);
+    //stop extra runs for enabled programs
+    pruneRuns(programRunsMap, programLatestRunMap);
+  }
+
+  private void gatherRunRecords(Map<ProgramId, List<ProgramRuntimeService.RuntimeInfo>> programRunsMap,
+                                Map<ProgramId, RunRecordDetail> programLastRunMap,
+                                ProgramRuntimeService.RuntimeInfo runtimeInfo, ProgramId programId) throws Exception {
+    programRunsMap.putIfAbsent(programId, new ArrayList<>());
+    programRunsMap.get(programId).add(runtimeInfo);
+    //Gather the latest run as we go
+    RunRecordDetail runRecordMeta = programLifecycleService
```

Comment:
Why need to load the run record? The runId has time encoded in it. You can simply order the runs by timestamp. See `RunIds.getTime()` method.

Reply:
Thanks, this makes it simpler.